pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -18,88 +18,128 @@ import (
+// The result contains a label "__tenant_id__" to identify the tenant ID that
 // it originally resulted from.
-// If the label tenantLabelName is already existing, its value is overwritten
+// If the label "__tenant_id__" is already existing, its value is overwritten
 // by the tenant ID and the previous value is exposed through a new label
-// prefixed with "original_". This behaviour is not implemented recursively
+// prefixed with "original_". This behaviour is not implemented recursively.
 func NewQueryable(upstream storage.Queryable) storage.Queryable {
-	return &mergeQueryable{
-		upstream: upstream,
+	return NewGenericQueryable(defaultTenantLabel, tenantQuerierCallback(upstream))
+}
+
+func tenantQuerierCallback(queryable storage.Queryable) GenericQuerierCallback {
+	return func(ctx context.Context, mint int64, maxt int64) ([]string, []storage.Querier, bool, error) {
+		tenantIDs, err := tenant.TenantIDs(ctx)
+		if err != nil {
+			return nil, nil, false, err
+		}
+
+		var queriers = make([]storage.Querier, len(tenantIDs))
+		for pos, tenantID := range tenantIDs {
+			q, err := queryable.Querier(
+				user.InjectOrgID(ctx, tenantID),
+				mint,
+				maxt,
+			)
+			if err != nil {
+				return nil, nil, false, err
+			}
+			queriers[pos] = q
+		}
+
+		// ensure to byPass a single querier and avoid adding a tenant label to
+		// ensure backward compatabiltiy when enabling multi-tenant query
+		// federation.
+		byPassWithSingleQuerier := true
+
+		return tenantIDs, queriers, byPassWithSingleQuerier, nil
+	}
+}
+
+// GenericQuerierCallback returns the underlying queriers and labelValues
+// relevant for the query. If the callback returns exactly a single querier,
+// byPassWithSingleQuerierUsing determines if it should still be using the
+// mergeQuerier (and adding a `identifyingLabelName` label) or if just the
+// underlying querier should be returned.
+type GenericQuerierCallback func(ctx context.Context, mint int64, maxt int64) (labelValues []string, queriers []storage.Querier, byPassWithSingleQuerier bool, err error)
+
+// NewGenericQueryable returns a queryable that merges results from multiple
+// underlying Queryables. The underlying queryables and its label values to be
+// considered are returned by a GenericQuerierCallback.
+// Results contain a label `identifyingLabelName` to identify the underlying
+// queryable that it originally resulted from.
+// If the label `identifyingLabelName` is already existing, its value is
+// overwritten and the previous value is exposed through a new label prefixed
+// with "original_". This behaviour is not implemented recursively.
+func NewGenericQueryable(identifyingLabelName string, callback GenericQuerierCallback) storage.Queryable {
+	return &genericQueryable{
+		identifyingLabelName: identifyingLabelName,
+		callback:             callback,
 	}
 }
 
-type mergeQueryable struct {
-	upstream storage.Queryable
+type genericQueryable struct {
```

Comment:
why rename it from `mergeQueryable`? Current name (`genericQueryable`) is too generic.

Reply:
Yes agree, naming is hard :grimacing: